timer/timer.go

```diff
@@ -0,0 +1,120 @@
+package timer
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/nivista/steady/.gen/protos/common"
+	"github.com/nivista/steady/internal/.gen/protos/messaging"
+)
+
+type (
+	// Timer represents the configuration for the execution of a recurring task.
+	Timer struct {
+		executer  executer
+		scheduler scheduler
+	}
+
+	// Progress records the progress of a recurring task.
+	Progress struct {
+		// CompletedExecutions indicated the number of successfully completed executions.
+		CompletedExecutions int
+
+		// LastExecution indicates which execution was recorded last.
+		LastExecution time.Time
+	}
+
+	executer interface {
+		execute()
+	}
+
+	scheduler interface {
+		schedule(prog Progress, now time.Time) (nextFire time.Time, done bool)
+	}
+)
+
+func getCanceller() (cancelled <-chan struct{}, cancel func()) {
+	var stopped int32
+
+	ch := make(chan struct{})
+
+	cancel = func() {
+		if atomic.CompareAndSwapInt32(&stopped, 0, 1) {
+			close(ch)
+		}
+	}
+
+	return ch, cancel
+}
+
+// Run starts the timers execution.
+func (t *Timer) Run(updateProgress func(Progress), finishTimer func(), initialProgress Progress, clock clockwork.Clock) (cancel func()) {
+
+	var progress = initialProgress
+
+	cancelled, cancel := getCanceller()
+
+	go func() {
+		for {
+			nextFire, done := t.scheduler.schedule(progress, clock.Now())
+
+			if done {
+				finishTimer()
+				cancel()
+				return
+			}
+
+			select {
+			case <-clock.After(nextFire.Sub(clock.Now())):
+				t.executer.execute()
+				progress.CompletedExecutions++
```

Comment:
here you are executing completed executions every time you call execute - but your comment says this represents successful execution. What happens in the case that you call execute and it returns an error?

Reply:
execute doesn't return an error. But you're right a web request could return an error. This brings up the question of retry behavior and stuff, I'll address this.